src/execution/operator/aggregate/physical_window.cpp

```diff
@@ -21,63 +22,434 @@ namespace duckdb {
+		payload_layout.Initialize(payload_types);
+		global_sort = make_unique<GlobalSortState>(buffer_manager, orders_copy, payload_layout);
+	}
+
+	void Sink(DataChunk &sort_buffer, DataChunk &payload_buffer) {
+		lock_guard<mutex> guard(lock);
+		if (!local_sort) {
+			local_sort = make_unique<LocalSortState>();
+			local_sort->Initialize(*global_sort, global_sort->buffer_manager);
+		}
+
+		local_sort->SinkChunk(sort_buffer, payload_buffer);
+	}
+
+	void Sort() {
+		if (local_sort) {
+			global_sort->AddLocalState(*local_sort);
+			local_sort.reset();
+		}
+
+		global_sort->PrepareMergePhase();
+		while (global_sort->sorted_blocks.size() > 1) {
+			global_sort->InitializeMergeRound();
+			MergeSorter merge_sorter(*global_sort, global_sort->buffer_manager);
+			merge_sorter.PerformInMergeRound();
+			global_sort->CompleteMergeRound(true);
+		}
+	}
+
+	Types sort_types;
+	mutex lock;
+	GlobalSortStatePtr global_sort;
+	LocalSortStatePtr local_sort;
+	idx_t memory_per_thread;
+
+private:
+};
+
 //	Global sink state
-class WindowGlobalState : public GlobalSinkState {
+class WindowGlobalSinkState : public GlobalSinkState {
 public:
-	WindowGlobalState(const PhysicalWindow &op_p, ClientContext &context)
-	    : op(op_p), buffer_manager(BufferManager::GetBufferManager(context)),
+	using HashGroupPtr = unique_ptr<WindowGlobalHashGroup>;
+	using Orders = vector<BoundOrderByNode>;
+	using Types = vector<LogicalType>;
+
+	WindowGlobalSinkState(const PhysicalWindow &op_p, ClientContext &context)
+	    : op(op_p), buffer_manager(BufferManager::GetBufferManager(context)), memory_per_thread(0),
 	      mode(DBConfig::GetConfig(context).window_mode) {
+
+		D_ASSERT(op.select_list[0]->GetExpressionClass() == ExpressionClass::BOUND_WINDOW);
+		auto wexpr = reinterpret_cast<BoundWindowExpression *>(op.select_list[0].get());
+
+		// we sort by both 1) partition by expression list and 2) order by expressions
+		payload_types = op.children[0]->types;
+
+		for (idx_t prt_idx = 0; prt_idx < wexpr->partitions.size(); prt_idx++) {
+			auto &pexpr = wexpr->partitions[prt_idx];
+			payload_types.push_back(pexpr->return_type);
+
+			if (wexpr->partitions_stats.empty() || !wexpr->partitions_stats[prt_idx]) {
+				orders.emplace_back(OrderType::ASCENDING, OrderByNullType::NULLS_FIRST, pexpr->Copy(), nullptr);
+			} else {
+				orders.emplace_back(OrderType::ASCENDING, OrderByNullType::NULLS_FIRST, pexpr->Copy(),
+				                    wexpr->partitions_stats[prt_idx]->Copy());
+			}
+		}
+
+		for (const auto &order : wexpr->orders) {
+			auto &oexpr = order.expression;
+			payload_types.push_back(oexpr->return_type);
+			orders.emplace_back(order.Copy());
+		}
+
+		// Memory usage per thread should scale with max mem / num threads
+		// We take 1/4th of this, to be conservative
+		idx_t max_memory = buffer_manager.GetMaxMemory();
+		idx_t num_threads = TaskScheduler::GetScheduler(context).NumberOfThreads();
+		memory_per_thread = (max_memory / num_threads) / 4;
```

Comment:
This seems to be copy pasted from `physical_order.cpp`, perhaps turn this into a method there? Especially since it is a slightly evil heuristic it would be good to have it in one place :)

Reply:
It's a bit worse than that... 

Since each thread can buffer up to 1024 sets of data, the local sorts should really be flushed with 1 / 1024 of that much memory. But the original heuristic should be shared.